admin/src/js/services/message-queue.js

```diff
@@ -0,0 +1,337 @@
+      var patientIds = compactUnique(messages.map(function(message) {
+        // don't process items which already have generated messages
+        return !message.sms && message.record && message.record.patient_id;
+      }));
+
+      if (!patientIds.length) {
+        return Promise.resolve(messages);
+      }
+
+      var referenceKeys = patientIds.map(function(patientId) {
+        return [ 'shortcode', patientId ];
+      });
+
+      return $q
+        .all([
+          DB({ remote: true }).query('medic-client/contacts_by_reference', { keys: referenceKeys }),
+          DB({ remote: true }).query('medic-client/registered_patients', { keys: patientIds, include_docs: true })
+        ])
+        .then(function(results) {
+          var contactsByReference = results[0];
+          var registrations = getValidRegistrations(results[1], settings);
+
+          messages.forEach(function(message) {
+            message.context = {
+              patient_uuid: findPatientUuid(contactsByReference, message),
+              registrations: findRegistrations(registrations, message)
+            };
+          });
+
+          return messages;
+        });
+    };
+
+    var hydrateContacts = function(messages) {
+      var contactIds = [];
+      messages.forEach(function(message) {
+        if (!message.sms) {
+          contactIds.push(message.context.patient_uuid, message.record.contact && message.record.contact._id);
+        }
+      });
+      contactIds = compactUnique(contactIds);
+      if (!contactIds.length) {
+        return messages;
+      }
+
+      return MessageQueueUtils.lineage
+        .fetchLineageByIds(contactIds)
+        .then(function(docList) {
+          var hydratedDocs = docList.map(function(docs) {
+            var doc = docs.shift();
+            return {
+              id: doc._id,
+              doc: MessageQueueUtils.lineage.fillParentsInDocs(doc, docs)
+            };
+          });
+
+          messages.forEach(function(message) {
+            message.context.patient = findContactById(hydratedDocs, message.context.patient_uuid);
+            message.contact = findContactById(hydratedDocs, message.record.contact && message.record.contact._id);
+          });
+
+          return messages;
+        });
+    };
+
+    var generateScheduledMessages = function(messages, settings) {
+      var translate = function(key, locale) {
+        return $translate.instant(key, null, 'no-interpolation', locale, null);
+      };
+
+      messages.forEach(function(message) {
+        if (message.sms) {
+          return;
+        }
+
+        var content = {
+          translationKey: message.scheduled_sms.translation_key,
+          message: message.scheduled_sms.content
+        };
+        var pseudoDoc = {
+          _id: message.record.id,
+          contact: message.contact,
+          fields: message.record.fields,
+          locale: message.record.locale
+        };
```

Comment:
I'm not sure about this pseudoDoc - theoretically the translation key may reference some field that you haven't copied, right? Can we include_docs and pass the real doc in instead?